### PR TITLE
Multiple Swift versions support

### PR DESCRIPTION
**Pull request by dnkoutso**
Requires https://github.com/CocoaPods/Core/pull/467
closes https://github.com/CocoaPods/CocoaPods/issues/8191

- [x] More Tests

Note: This PR does not add Swift variants (i.e splitting pods across Swift versions) it maintains the status quo and ensures validation occurs.

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Multiple Swift versions support  
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: spec/unit/installer/xcode/target_validator_spec.rb [hunk 1]
```diff
@@ -273,17 +273,16 @@ def create_validator(sandbox, podfile, lockfile, integrate_targets = false)
               use_frameworks!
               pod 'OrangeFramework', :path => (fixture_path + 'orange-framework').to_s
               pod 'matryoshka',      :path => (fixture_path + 'matryoshka').to_s
-              target 'SampleProject'
```
Comment (dnkoutso) [inline spec/unit/installer/xcode/target_validator_spec.rb]: updated most of these tests *not* to use stabbing. Much better imo. Added tests for the new logic here too.

File: lib/cocoapods/project.rb [hunk 2]
```diff
@@ -383,7 +381,7 @@ def group_for_path_in_group(absolute_pathname, group, reflect_file_system_struct
     # Returns the name to be used for a the variant group for a file at a given path.
     # The path must be localized (within an *.lproj directory).
     #
-    # @param  [Pathname] The localized path to get a variant group name for.
+    # @param  [Pathname] path The localized path to get a variant group name for.
```
Comment (amorde) [inline lib/cocoapods/project.rb]: haha thank you - this has been bugging me but I never found a good time to update it

File: lib/cocoapods/target/pod_target.rb [hunk 3]
```diff
@@ -127,18 +128,28 @@ def label
       end
     end
 
-    # @return [String] the Swift version for the target. If the pod author has provided a Swift version
-    #                  then that is the one returned, otherwise the Swift version is determined by the user
-    #                  targets that include this pod target.
-    #
+    # @return [String] the Swift version for the target. If the pod author has provided a set of Swift versions
+    #         supported by their pod then the max Swift version across all of target definition is chosen, unless
```
Comment (amorde) [inline lib/cocoapods/target/pod_target.rb]: nit: target definition => target definition**s**

File: lib/cocoapods/target/pod_target.rb [hunk 4]
```diff
@@ -127,18 +128,28 @@ def label
...
+    #         a target definition specifies explicit requirements for supported Swift versions. Otherwise the Swift
+    #         version is derived by the target definitions that integrate this pod as long as its a single one.
```
Comment (amorde) [inline lib/cocoapods/target/pod_target.rb]: `as long as they are the same`?

File: lib/cocoapods/target/pod_target.rb [hunk 5]
```diff
@@ -328,7 +339,7 @@ def product_type_for_test_type(test_type)
       when :unit
         :unit_test_bundle
       else
-        raise Informative, "Unknown test type `#{test_type}`."
+        raise ArgumentError, "Unknown test type `#{test_type}`."
```
Comment (amorde) [inline lib/cocoapods/target/pod_target.rb]: nice

File: lib/cocoapods/validator.rb [hunk 6]
```diff
@@ -322,6 +312,24 @@ def dot_swift_version
       swift_version_path.read.strip
     end
 
+    # @return [String] The derived Swift version to use for validation. The order of precedence is as follows:
+    #         - The `--swift-version` parameter is always checked first and honored if passed.
+    #         - The `swift_versions` attribute within in the podspec, in which case the latest version is always chosen.
```
Comment (amorde) [inline lib/cocoapods/validator.rb]: `within the podspec`

File: spec/unit/installer/xcode/target_validator_spec.rb (hunk 1 above)
Comment (amorde) [reply spec/unit/installer/xcode/target_validator_spec.rb]: much much cleaner IMO

**Comment by amorde** [comment]
haven't looked too deeply into the tests yet but overall looks great to me

File: spec/unit/installer/xcode/target_validator_spec.rb [hunk 7]
```diff
@@ -306,22 +305,17 @@ def create_validator(sandbox, podfile, lockfile, integrate_targets = false)
...
+            @validator = create_validator(config.sandbox, podfile, lockfile)
             e = should.raise Informative do
               @validator.validate!
             end
             e.message.should.match /Unable to determine Swift version for the following pods:/
             e.message.should.include '`OrangeFramework` does not specify a Swift version and none of the targets ' \
               '(`SampleProject` and `TestRunner`) integrating it have the `SWIFT_VERSION` attribute set. Please contact ' \
               'the author or set the `SWIFT_VERSION` attribute in at least one of the targets that integrate this pod.'
-            e.message.should.include '`matryoshka` does not specify a Swift version and none of the targets ' \
+            e.message.should.not.include '`matryoshka` does not specify a Swift version and none of the targets ' \
```
Comment (dnkoutso) [inline spec/unit/installer/xcode/target_validator_spec.rb]: this is now switched because the real podspec does not include swift.

File: spec/unit/target/pod_target_spec.rb [hunk 8]
```diff
@@ -122,14 +121,49 @@ module Pod
...
+        target_definition_two = fixture_target_definition('App2')
+        target_definition_two.store_swift_version_requirements('= 4.2')
+        pod_target = PodTarget.new(config.sandbox, false, {}, [], Platform.ios, [@banana_spec], [target_definition_one,
+                                                                                                 target_definition_two])
+        @pod_target.root_spec.stubs(:swift_versions).returns([Version.new('3.0'), Version.new('4.0'),
+                                                              Version.new('4.2')])
+        pod_target.swift_version.should == '4.2'
+        end
+
+      it 'returns the swift version with the given requirements from the target definition' do
```
Comment (dnkoutso) [inline spec/unit/target/pod_target_spec.rb]: dup test name will fix

File: spec/unit/validator_spec.rb [hunk 9]
```diff
@@ -1030,137 +1031,138 @@ def test_swiftpod_with_dot_swift_version(version = '3.1.0')
...
-          validator.results.count.should == 0
-        end
+        validator = test_swiftpod('3.1.0')
+        validator.validate
+        validator.results.count.should == 0
+      end
 
-        it 'succeeds with a .swift-version file' do
-          Specification.any_instance.stubs(:deployment_target).returns('9.0')
+      it 'succeeds with a .swift-version file' do
```
Comment (dnkoutso) [inline spec/unit/validator_spec.rb]: here I fixed indentation problem

File: lib/cocoapods/validator.rb [hunk 10]
```diff
@@ -322,6 +313,24 @@ def dot_swift_version
...
+    #         - The `--swift-version` parameter is always checked first and honored if passed.
+    #         - The `swift_versions` DSL attribute within the podspec, in which case the latest version is always chosen.
+    #         - The Swift version within the `.swift-version` file if present.
+    #         - If none of the above are set then the `#DEFAULT_SWIFT_VERSION` is used.
+    #
+    def derived_swift_version
+      @derived_swift_version ||= begin
+        if !swift_version.nil?
+          swift_version
+        elsif (version = spec.swift_versions.max) || (version = dot_swift_version)
```
[the remark this answers is not in the document]
Comment (dnkoutso) [reply lib/cocoapods/validator.rb]: fixed

File: lib/cocoapods/target/pod_target.rb [hunk 11]
```diff
@@ -127,18 +128,30 @@ def label
...
+    #         a target definition specifies explicit requirements for supported Swift versions. Otherwise the Swift
+    #         version is derived by the target definitions that integrate this pod as long as they are the same.
     #
     def swift_version
-      spec_swift_version || target_definitions.map(&:swift_version).compact.uniq.first
+      @swift_version ||= begin
+        if spec_swift_versions.empty?
+          target_definitions.map(&:swift_version).compact.uniq.first
+        else
+          spec_swift_versions.select do |swift_version|
```
[the remark this answers is not in the document]
Comment (dnkoutso) [reply lib/cocoapods/target/pod_target.rb]: updated

File: lib/cocoapods/target/pod_target.rb [hunk 12]
```diff
@@ -127,18 +128,30 @@ def label
...
+    # @return [String] the Swift version for the target. If the pod author has provided a set of Swift versions
+    #         supported by their pod then the max Swift version across all of target definitions is chosen, unless
+    #         a target definition specifies explicit requirements for supported Swift versions. Otherwise the Swift
+    #         version is derived by the target definitions that integrate this pod as long as they are the same.
     #
     def swift_version
-      spec_swift_version || target_definitions.map(&:swift_version).compact.uniq.first
+      @swift_version ||= begin
+        if spec_swift_versions.empty?
+          target_definitions.map(&:swift_version).compact.uniq.first
```
Comment (segiddins) [inline lib/cocoapods/target/pod_target.rb]: `target_definitions.map(&:swift_version).find`

File: CHANGELOG.md [hunk 13]
```diff
@@ -8,6 +8,14 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Enhancements
 
+* Multiple Swift versions support  
+  [Dimitris Koutsogiorgas](https://github.com/dnkoutso)
+  [#8191](https://github.com/CocoaPods/CocoaPods/issues/8191)
+
+* Adds app spec project integration.
```
Comment (dnkoutso) [inline CHANGELOG.md]: this was wrongly placed in 1.6.0.beta.2 section. Moved it in a separate commit as part of this PR

File: lib/cocoapods/target/pod_target.rb [hunk 14]
```diff
@@ -127,18 +128,30 @@ def label
...
+        else
+          spec_swift_versions.sort.reverse_each.find do |swift_version|
+            target_definitions.all? do |td|
+              td.supports_swift_version?(swift_version)
+            end
+          end.to_s
```
Comment (amorde) [inline lib/cocoapods/target/pod_target.rb]: out of scope of this PR but eventually this should probably transition to some sort of `SwiftVersion` (or `Version`) type, especially since `swift_versions` returns `Array<Version>`

Comment (dnkoutso) [reply lib/cocoapods/target/pod_target.rb]: hm I can do in a follow up for `Version`
